subsys/storage/storage_area/storage_area_store.c

```diff
@@ -0,0 +1,1155 @@
+/*
+ * Copyright (c) 2024 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/crc.h>
+#include <zephyr/storage/storage_area/storage_area_store.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(storage_area_store, CONFIG_STORAGE_AREA_LOG_LEVEL);
+
+/* record magic: chosen to be different from the erase-value (0x00 or 0xFF) */
+#define SAS_MAGIC      0xF0
+#define SAS_FILLVAL    0xFF
+
+/* header size: record magic (1 BYTE) + wrapcnt (1 BYTE) + size (2 BYTE) */
+#define SAS_HDRSIZE    4
+#define SAS_CRCINIT    0
+#define SAS_CRCSIZE    sizeof(uint32_t)
+#define SAS_MINBUFSIZE 32
+
+#define SAS_MIN(a, b)             (a < b ? a : b)
+#define SAS_MAX(a, b)             (a < b ? b : a)
+#define SAS_ALIGNUP(num, align)   (((num) + ((align) - 1)) & ~((align) - 1))
+#define SAS_ALIGNDOWN(num, align) ((num) & ~((align) - 1))
+
+static void sector_advance(const struct storage_area_store *store,
+			   size_t *sector, size_t cnt)
+{
+	while (cnt > 0U) {
+		(*sector)++;
+		cnt--;
+		if ((*sector) == store->sector_cnt) {
+			*sector = 0U;
+		}
+	}
+}
+
+static ALWAYS_INLINE int
+store_init_semaphore(const struct storage_area_store *store)
+{
+#ifdef CONFIG_STORAGE_AREA_STORE_SEMAPHORE
+	return k_sem_init(&store->data->semaphore, 1, 1);
+#else
+	ARG_UNUSED(store);
+	return 0;
+#endif /* CONFIG_STORAGE_AREA_STORE_SEMAPHORE */
+}
+
+static ALWAYS_INLINE int
+store_take_semaphore(const struct storage_area_store *store)
+{
+#ifdef CONFIG_STORAGE_AREA_STORE_SEMAPHORE
+	return k_sem_take(&store->data->semaphore, K_FOREVER);
+#else
+	ARG_UNUSED(store);
+	return 0;
+#endif /* CONFIG_STORAGE_AREA_STORE_SEMAPHORE */
+}
+
+static ALWAYS_INLINE void
+store_give_semaphore(const struct storage_area_store *store)
+{
+#ifdef CONFIG_STORAGE_AREA_STORE_SEMAPHORE
+	(void)k_sem_give(&store->data->semaphore);
```

Comment:
Return type is `void`

```suggestion
	k_sem_give(&store->data->semaphore);
```